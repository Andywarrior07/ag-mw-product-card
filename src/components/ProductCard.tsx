import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  ProductContextProps,
  Product,
  OnChangeArgs,
  InitialValues,
  ProductCardHandlers,
} from '../interfaces/Product.interface';

import styles from '../styles/styles.module.css';

interface Props {
  product: Product;
  className?: string;
  initialValues?: InitialValues;
  value?: number;
  onChange?: (args: OnChangeArgs) => void;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  style?: CSSProperties;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  product,
  className,
  initialValues,
  value,
  onChange,
  style,
  children,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ counter, product, increaseBy, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          reset,
          product,
          increaseBy,
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
        } as ProductCardHandlers)}
      </div>
    </Provider>
  );
};
