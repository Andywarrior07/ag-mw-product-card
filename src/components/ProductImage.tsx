import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';

import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const imgToShow = img ? img : product.img ? product.img : noImage;

  return (
    <img
      src={imgToShow}
      alt="Product pic"
      style={style}
      className={`${styles.productImg} ${className}`}
    />
  );
};
