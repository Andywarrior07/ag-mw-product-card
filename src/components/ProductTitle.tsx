import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
  title?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: Props) => {
  const { product } = useContext(ProductContext);
  const titleText = title ? title : product.title;

  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {titleText}
    </span>
  );
};
