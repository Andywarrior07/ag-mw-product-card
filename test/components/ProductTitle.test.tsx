import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('should render component correctly', () => {
    const title = 'Test title';

    const wrapper = renderer.create(<ProductTitle title={title} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should show name of the product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
