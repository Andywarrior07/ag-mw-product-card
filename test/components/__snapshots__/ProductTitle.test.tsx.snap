// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should render component correctly 1`] = `
<span
  className="productDescription undefined"
>
  Test title
</span>
`;

exports[`ProductTitle should show name of the product 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Without image
  </span>
</div>
`;
